src/main/java/org/dataone/hashstore/hashstoreconverter/FileHashStoreLinks.java

```diff
@@ -0,0 +1,233 @@
+            // Create parent directories to the hard link, otherwise
+            // Files.createLink will throw a NoSuchFileException
+            FileHashStoreUtility.createParentDirectories(objHardLinkPath);
+
+            try {
+                Files.createLink(objHardLinkPath, filePath);
+
+            } catch (FileAlreadyExistsException faee) {
+                logFileHashStoreLinks.warn("Data object already exists at: " + objHardLinkPath);
+            }
+
+            // This method is thread safe and synchronized
+            tagObject(pid, objectCid);
+            logFileHashStoreLinks.info(
+                "Hard link has been created for pid:" + pid + " with cid: " + objectCid
+                    + ", and has been tagged");
+
+            return new ObjectMetadata(pid, objectCid, Files.size(objHardLinkPath), hexDigests);
+
+        } finally {
+            // Close stream
+            fileStream.close();
+        }
+    }
+
+    /**
+     * Get a HashStore data object path
+     *
+     * @param pid Persistent or authority-based identifier
+     * @return Path to a HashStore data object
+     * @throws NoSuchAlgorithmException Conflicting algorithm preventing calculation of the path
+     * @throws IOException              If there is an issue with reading from the pid refs file
+     */
+    protected Path getHashStoreLinksDataObjectPath(String pid)
+        throws NoSuchAlgorithmException, IOException {
+        return getHashStoreDataObjectPath(pid);
+    }
+
+    /**
+     * Get a HashMap consisting of algorithms and their respective hex digests for a given data
+     * stream. If an additional algorithm is supplied and supported, it and its checksum value will
+     * be included in the hex digests map. Default algorithms: MD5, SHA-1, SHA-256, SHA-384,
+     * SHA-512
+     *
+     * @param dataStream          input stream of data to store
+     * @param additionalAlgorithm additional algorithm to include in hex digest map
+     * @return A map containing the hex digests of the default algorithms
+     * @throws NoSuchAlgorithmException Unable to generate new instance of supplied algorithm
+     * @throws IOException              Issue with writing file from InputStream
+     * @throws SecurityException        Unable to write to tmpFile
+     * @throws FileNotFoundException    tmpFile cannot be found
+     */
+    protected Map<String, String> generateChecksums(
+        InputStream dataStream, String additionalAlgorithm)
+        throws NoSuchAlgorithmException, IOException, SecurityException {
+        // Determine whether to calculate additional or checksum algorithms
+        boolean generateAddAlgo = false;
+        if (additionalAlgorithm != null) {
+            validateAlgorithm(additionalAlgorithm);
+            generateAddAlgo = shouldCalculateAlgorithm(additionalAlgorithm);
+        }
+
+        MessageDigest md5 = MessageDigest.getInstance(DefaultHashAlgorithms.MD5.getName());
```

Comment:
This a low priority issue. In the FileHashStore.writeToTmpFileAndGenerateChecksums method, it has the same hard-coded to create MessageDigest. So if we need to change default checksum algorithms, we have two places to take care of. Is it possible to store the default checksum algorithm in a list and both code just loop the variable?

Reply:
Thank you for bringing this up! I have refactored both `FileHashStore` and `FileHashStoreLinks` to calculate checksums based on the `DefaultHashAlgorithms` enum object. So all we have to do in the future should we desire a change is update the enum object.